discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -0,0 +1,324 @@
+import concurrent.futures
+import datetime
+import logging
+import re
+import time
+from typing import Tuple, Optional
+from sqlalchemy import desc, and_
+from solana.publickey import PublicKey
+from src.tasks.celery_app import celery
+from src.utils.config import shared_config
+from src.utils.solana import get_address_pair
+from src.models import User, UserBankTransaction, UserBankAccount
+from src.queries.get_balances import enqueue_balance_refresh
+
+logger = logging.getLogger(__name__)
+
+# Populate values used in UserBank indexing from config
+USER_BANK_ADDRESS = shared_config["solana"]["user_bank_program_address"]
+WAUDIO_PROGRAM_ADDRESS = shared_config["solana"]["waudio_program_address"]
+USER_BANK_KEY = PublicKey(USER_BANK_ADDRESS) if USER_BANK_ADDRESS else None
+WAUDIO_PROGRAM_PUBKEY = (
+    PublicKey(WAUDIO_PROGRAM_ADDRESS) if WAUDIO_PROGRAM_ADDRESS else None
+)
+
+# Static SPL Token Program ID
+SPL_TOKEN_ID = PublicKey("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA")
+
+# Used to limit tx history if needed
+MIN_SLOT = int(shared_config["solana"]["user_bank_min_slot"])
+
+# Maximum number of batches to process at once
+TX_SIGNATURES_MAX_BATCHES = 3
+
+# Last N entries present in tx_signatures array during processing
+TX_SIGNATURES_RESIZE_LENGTH = 3
+
+# Recover ethereum public key from bytes array
+# Message formatted as follows:
+# EthereumAddress = [214, 237, 135, 129, 143, 240, 221, 138, 97, 84, 199, 236, 234, 175, 81, 23, 114, 209, 118, 39]
+def parse_eth_address_from_msg(msg):
+    logger.error(f"index_user_bank.py {msg}")
+    res = re.findall(r"\[.*?\]", msg)
+    # Remove brackets
+    inner_res = res[0][1:-1]
+    # Convert to public key hex for ethereum
+    arr_val = [int(s) for s in inner_res.split(",")]
+    public_key_bytes = bytes(arr_val)
+    public_key = public_key_bytes.hex()
+    public_key_str = f"0x{public_key}"
+    return public_key_str, public_key_bytes
+
+
+# Return highest user bank slot that has been processed
+def get_highest_user_bank_tx_slot(db):
+    slot = MIN_SLOT
+    with db.scoped_session() as session:
+        tx_query = (
+            session.query(UserBankTransaction).order_by(desc(UserBankTransaction.slot))
+        ).first()
+        if tx_query:
+            slot = tx_query.slot
+    return slot
+
+
+# Query a tx signature and confirm its existence
+def get_tx_in_db(session, tx_sig):
+    exists = False
+    tx_sig_db_count = (
+        session.query(UserBankTransaction).filter(
+            UserBankTransaction.signature == tx_sig
+        )
+    ).count()
+    exists = tx_sig_db_count > 0
+    # logger.info(f"index_user_bank.py | {tx_sig} exists={exists}")
+    return exists
+
+
+# Retry 5x until a tx 'result' is found with valid contents
+# If not found, move forward
+def get_sol_tx_info(solana_client, tx_sig):
```

Comment:
i think we can actually import this from the track listen count task or make a common function